internal/venice-client-common/src/main/java/com/linkedin/venice/stats/metrics/MetricEntityStateTwoEnums.java

```diff
@@ -0,0 +1,179 @@
+        enumTypeClass2);
+  }
+
+  /** Overloaded Factory method for constructor with Tehuti parameters */
+  public static <E1 extends Enum<E1> & VeniceDimensionInterface, E2 extends Enum<E2> & VeniceDimensionInterface> MetricEntityStateTwoEnums<E1, E2> create(
+      MetricEntity metricEntity,
+      VeniceOpenTelemetryMetricsRepository otelRepository,
+      TehutiSensorRegistrationFunction registerTehutiSensorFn,
+      TehutiMetricNameEnum tehutiMetricNameEnum,
+      List<MeasurableStat> tehutiMetricStats,
+      Map<VeniceMetricsDimensions, String> baseDimensionsMap,
+      Class<E1> enumTypeClass1,
+      Class<E2> enumTypeClass2) {
+    return new MetricEntityStateTwoEnums<>(
+        metricEntity,
+        otelRepository,
+        registerTehutiSensorFn,
+        tehutiMetricNameEnum,
+        tehutiMetricStats,
+        baseDimensionsMap,
+        enumTypeClass1,
+        enumTypeClass2);
+  }
+
+  /**
+    * Creates an EnumMap of {@link Attributes} for each possible value of the dynamic dimensions
+   * {@link #enumTypeClass1} and {@link #enumTypeClass2}
+   */
+  private EnumMap<E1, EnumMap<E2, Attributes>> createAttributesEnumMap(
+      MetricEntity metricEntity,
+      VeniceOpenTelemetryMetricsRepository otelRepository,
+      Map<VeniceMetricsDimensions, String> baseDimensionsMap) {
+    if (!emitOpenTelemetryMetrics()) {
+      return null;
+    }
+
+    EnumMap<E1, EnumMap<E2, Attributes>> attributesEnumMap = new EnumMap<>(enumTypeClass1);
+    Map<VeniceMetricsDimensions, String> additionalDimensionsMap = new HashMap<>();
+    for (E1 enumConst1: enumTypeClass1.getEnumConstants()) {
+      additionalDimensionsMap.put(enumConst1.getDimensionName(), enumConst1.getDimensionValue());
+      EnumMap<E2, Attributes> mapE2 = new EnumMap<>(enumTypeClass2);
+      attributesEnumMap.put(enumConst1, mapE2);
+      for (E2 enumConst2: enumTypeClass2.getEnumConstants()) {
+        additionalDimensionsMap.put(enumConst2.getDimensionName(), enumConst2.getDimensionValue());
+        mapE2
+            .put(enumConst2, otelRepository.createAttributes(metricEntity, baseDimensionsMap, additionalDimensionsMap));
+      }
+    }
+    return attributesEnumMap;
+  }
+
+  Attributes getAttributes(E1 key1, E2 key2) {
+    if (!emitOpenTelemetryMetrics()) {
+      return null;
+    }
+    if (key1 == null || key2 == null) {
+      throw new IllegalArgumentException(
+          "The key for otel dimension cannot be null for metric Entity: " + getMetricEntity().getMetricName());
+    }
+    if (!enumTypeClass1.isInstance(key1) || !enumTypeClass2.isInstance(key2)) {
+      // defensive check: This can only happen if the instance is declared without the explicit types
+      // and passed in wrong args
+      throw new IllegalArgumentException(
+          "The key for otel dimension is not of the correct type: " + key1.getClass() + "," + key2.getClass()
+              + " for metric Entity: " + getMetricEntity().getMetricName());
+    }
```

Comment:
What does it mean that "the instance is declared without the explicit types"? Do we need to care?

Reply:
So what would be the runtime error for case 2? `ClassCastException`?

Reply:
I guess I would suggest the following:

Look up in the `EnumMap` directly. Then do the `if (attributes == null)` check you currently have at the end. Only if that is true, then do the other checks (null key, wrong class) in order to provide the clearest possible error message.

There is no point to pre-emptively check all of these things which the `EnumMap` already checks on your behalf later. In general, we want to optimize for the correct path, then add extra overhead only in the error handling path.